drivers/usb/udc/udc_mcux.c

```diff
@@ -0,0 +1,903 @@
+			}
+
+			udc_ep_buf_set_setup(buf);
+			memcpy(buf->data, mcux_msg->buffer, 8);
+			net_buf_add(buf, 8);
+
+			setup = (struct usb_setup_packet *)mcux_msg->buffer;
+			if ((setup->RequestType.type == USB_REQTYPE_TYPE_STANDARD) &&
+				(!setup->wLength) &&
+				(!(((setup->bmRequestType) >> 7) & 0x01U)) &&
+				(setup->bRequest == USB_SREQ_SET_ADDRESS)) {
+				usb_mcux_control(dev, kUSB_DeviceControlPreSetDeviceAddress,
+					&setup->wValue);
+			}
+
+			/* Update to next stage of control transfer */
+			udc_ctrl_update_stage(dev, buf);
+
+			if (!buf->len) {
+				return kStatus_USB_Error;
+			}
+
+			if (udc_ctrl_stage_is_data_out(dev)) {
+				/*  Allocate and feed buffer for data OUT stage */
+				LOG_DBG("s:%p|feed for -out-", buf);
+				err = usb_mcux_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+				if (err == -ENOMEM) {
+					err = udc_submit_ep_event(dev, buf, err);
+				}
+			} else if (udc_ctrl_stage_is_data_in(dev)) {
+				err = udc_ctrl_submit_s_in_status(dev);
+			} else {
+				err = udc_ctrl_submit_s_status(dev);
+			}
+		} else if (ep == USB_CONTROL_EP_IN) {
+			len = MIN(buf->len, mcux_msg->length);
+			buf->data += len;
+			buf->len -= len;
+#if defined(CONFIG_NOCACHE_MEMORY)
+			udc_mcux_nocache_free(mcux_msg->buffer);
+#endif
+
+			if (udc_ctrl_stage_is_status_in(dev) ||
+			udc_ctrl_stage_is_no_data(dev)) {
+				/* Status stage finished, notify upper layer */
+				udc_ctrl_submit_status(dev, buf);
+			}
+
+			/* Update to next stage of control transfer */
+			udc_ctrl_update_stage(dev, buf);
+
+			if (udc_ctrl_stage_is_status_out(dev)) {
+				/*
+				 * IN transfer finished, release buffer,
+				 * control OUT buffer should be already fed.
+				 */
+				net_buf_unref(buf);
+				err = usb_mcux_ctrl_feed_dout(dev, 0u);
+			}
+		} else if (ep == USB_CONTROL_EP_OUT) {
+			LOG_DBG("DataOut ep 0x%02x",  ep);
+			len = MIN(net_buf_tailroom(buf), mcux_msg->length);
+#if defined(CONFIG_NOCACHE_MEMORY)
+			memcpy(net_buf_tail(buf), mcux_msg->buffer, len);
+			udc_mcux_nocache_free(mcux_msg->buffer);
+#endif
+			net_buf_add(buf, len);
+			if (udc_ctrl_stage_is_status_out(dev)) {
+				/* Update to next stage of control transfer */
+				udc_ctrl_update_stage(dev, buf);
+				/* Status stage finished, notify upper layer */
+				udc_ctrl_submit_status(dev, buf);
+			} else {
+				/* Update to next stage of control transfer */
+				udc_ctrl_update_stage(dev, buf);
+			}
+
+			if (udc_ctrl_stage_is_status_in(dev)) {
+				err = udc_ctrl_submit_s_out_status(dev, buf);
+			}
+		} else {
+			if (USB_EP_DIR_IS_IN(ep)) {
+				len = MIN(buf->len, mcux_msg->length);
+				buf->data += len;
+				buf->len -= len;
+			} else {
+				len = MIN(net_buf_tailroom(buf), mcux_msg->length);
+#if defined(CONFIG_NOCACHE_MEMORY)
+				memcpy(net_buf_tail(buf), mcux_msg->buffer, len);
+#endif
+				net_buf_add(buf, len);
+			}
+
+#if defined(CONFIG_NOCACHE_MEMORY)
+			udc_mcux_nocache_free(mcux_msg->buffer);
+#endif
+			err = udc_submit_ep_event(dev, buf, 0);
```

Comment:
Perhaps you need to check ZLP flag(`udc_ep_buf_has_zlp(buf)`) and add ZLP transaction somewhere here.